api/types/registry/authconfig.go

```diff
@@ -33,6 +33,19 @@ type AuthConfig struct {
 	RegistryToken string `json:"registrytoken,omitempty"`
 }
 
+// EncodeAuthConfig serializes the auth configuration as a base64url encoded
+// RFC4648, section 5) JSON string for sending through the X-Registry-Auth header.
+//
+// For details on base64url encoding, see:
+// - RFC4648, section 5:   https://tools.ietf.org/html/rfc4648#section-5
+func EncodeAuthConfig(authConfig AuthConfig) (string, error) {
```

Comment:
Is the CLI code the intended consumer of this function?

Reply:
Currently, yes; we can use this function to replace the `` that's currently living in docker/cli;
https://github.com/docker/cli/blob/1f4111d2bf3dbd9b6f5252c551621af84401b691/cli/command/registry.go#L30-L37 

While admittedly, that'd mostly be shifting some code around, I think it makes sense to have the "encoding" and "decoding" logic to be together, and (as it's effectively part of the API's contract in what format to send this information), I think it makes sense to make it part of the API packages (instead of in "an" implementation of a CLI that uses the API).


w.r.t. the "currently" bit in the above; I'm somewhat learning towards making the encoding/decoding part of the _client_, so that code like https://github.com/docker/cli/blob/f1615facb1ca44e4336ab20e621315fc2cfb845a/cli/command/container/create.go#L127-L136 wouldn't be needed, and the "as-is" `AuthConfig` (or perhaps some other construct) could be passed as option.

I haven't worked on that yet (this would require changing some function signatures).